source/Octopus.Tentacle/Scripts/ScriptStateWorkspace.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Threading;
+using Newtonsoft.Json;
+using Octopus.Tentacle.Contracts;
+using Octopus.Tentacle.Util;
+
+namespace Octopus.Tentacle.Scripts
+{
+    internal class ScriptStateWorkspace : IScriptStateWorkspace
+    {
+        static readonly SemaphoreSlim ScriptStateLock = new(1, 1);
+
+        readonly IScriptWorkspace workspace;
+        readonly ScriptTicket scriptTicket;
+        readonly IOctopusFileSystem fileSystem;
+
+        public ScriptStateWorkspace(ScriptTicket scriptTicket, IScriptWorkspace workspace, IOctopusFileSystem fileSystem)
+        {
+            this.scriptTicket = scriptTicket;
+            this.workspace = workspace;
+            this.fileSystem = fileSystem;
+        }
+
+        string StateFilePath => workspace.ResolvePath("scriptstate.json");
+
+        public bool StateExists => Exists(aquireLock: true);
```

Comment:
The Exists method could be renamed `ExistsNoLock()` (Dropping the param) instead since it seems every other calls to exists takes the lock itself. This place this the only place that doesn't take a lock.